thrall/test/lib/kinesis/MessageProcessorTest.scala

```diff
@@ -0,0 +1,89 @@
+package lib.kinesis
+
+import java.util.UUID
+
+import com.gu.mediaservice.lib.aws.{BulkIndexRequest, UpdateMessage}
+import com.gu.mediaservice.model.leases.MediaLease
+import com.gu.mediaservice.model.usage.UsageNotice
+import com.gu.mediaservice.model.{Collection, Crop, Edits, Image, ImageMetadata, SyndicationRights}
+import lib.{BulkIndexS3Client, MetadataEditorNotifications, ThrallStore}
+import lib.elasticsearch.{ElasticSearchTestBase, SyndicationRightsOps}
+import org.joda.time.DateTime
+import org.scalatest.mockito.MockitoSugar
+import play.api.libs.json.{JsArray, Json}
+
+import scala.concurrent.{Await, Future}
+
+
+class MessageProcessorTest extends ElasticSearchTestBase with MockitoSugar {
+  "MessageProcessor" - {
+    val messageProcessor = new MessageProcessor(
+      es = ES,
+      store = mock[ThrallStore],
+      metadataEditorNotifications = mock[MetadataEditorNotifications],
+      syndicationRightsOps = mock[SyndicationRightsOps],
+      bulkIndexS3Client = mock[BulkIndexS3Client])
+
+    "usages" - {
+
+      "adds usages" - {
+        "for an image that exists" in {
+
+          val id = UUID.randomUUID().toString
+
+          val userMetadata = Some(Edits(metadata = ImageMetadata(
+            description = Some("My boring image"),
+            title = Some("User supplied title"),
+            subjects = List("foo", "bar"),
+            specialInstructions = Some("Testing")
+          )))
+
+          val image = createImageForSyndication(id = UUID.randomUUID().toString, true, Some(DateTime.now()), None).
+            copy(userMetadata = userMetadata)
+
+          Await.result(Future.sequence(ES.indexImage(id, Json.toJson(image))), fiveSeconds)
+
+          eventually(timeout(fiveSeconds), interval(oneHundredMilliseconds))(reloadedImage(id).map(_.id) shouldBe Some(image.id))
+
+          val message = UpdateMessage(
+            "update-image-usages",
+            image = None,
+            id = Some(id),
+            usageNotice = Some(UsageNotice(id, JsArray())),
+            edits = None,
+            lastModified = Some(new DateTime()),
+            collections = None,
+            leaseId = None,
+            crops = None,
+            mediaLease = None,
+            leases = None,
+            syndicationRights = None,
+            bulkIndexRequest = None
+          )
+          noException should be thrownBy Await.result(messageProcessor.updateImageUsages(message), fiveSeconds)
+        }
+        "not crash for an image that doesn't exist 👻🖼" in {
```

Comment:
I'm not sure this test is exhaustive. IIUC this PR is designed to fail fast on a 404, that is, do not retry the message more than once. This test is only asserting that an exception isn't thrown. Could we add a test against the retry behaviour in a 404 scenario too please?

Reply:
Adding tests to the retry handler should be its own pr.

Reply:
Ok. How can we determine success of this change? The current structure may make adding this type of test obscure; can we add detailed logging which we can observe? Logging the attempt number with the subject, for example? At the moment, it's not clear that this PR solves the issue we've witnessed where a message gets retried and blocks the remaining messages.

Reply:
We found one of the root causes being retries on missing image usages. This makes those changes return a success instead of a failure state- and tests for this. 

This PR also makes that information available in the response from executeAndLog and increases of the specificity of logging. 

Extra work is needed in:
- some tests for the retry handler
- better handling of logmarkers to make this more explicit 
- using the api change in this PR to solve the syndication form of this bug

I don't think this PR is enough to fix the issue, but imo everything that is, has been blocked by this PR for the past two weeks.

Reply:
> This makes those changes return a success instead of a failure state- and tests for this.

Can this test test for that? As in, rather than asserting no exception was thrown, can we assert on the return type? IIRC, we retry iff return isn't `Success` (and not timeout). Asserting for `Success` would provide more confidence that the message won't get retried.

Reply:
Tests modified to check for Success.